src/test/java/org/apache/commons/collections4/bloomfilter/LayerManagerTest.java

```diff
@@ -0,0 +1,305 @@
+
+    private Shape shape = Shape.fromKM(17, 72);
+
+    private LayerManager.Builder testingBuilder() {
+        return LayerManager.builder().setSupplier(() -> new SimpleBloomFilter(shape));
+    }
+
+    @Test
+    public void testAdvanceOnPopulated() {
+        Predicate<LayerManager> underTest = LayerManager.ExtendCheck.advanceOnPopulated();
+        LayerManager layerManager = testingBuilder().build();
+        assertFalse(underTest.test(layerManager));
+        layerManager.getTarget().merge(TestingHashers.FROM1);
+        assertTrue(underTest.test(layerManager));
+    }
+
+    @Test
+    public void testNeverAdvance() {
+        Predicate<LayerManager> underTest = LayerManager.ExtendCheck.neverAdvance();
+        LayerManager layerManager = testingBuilder().build();
+        assertFalse(underTest.test(layerManager));
+        for (int i = 0; i < 10; i++) {
+            layerManager.getTarget().merge(TestingHashers.randomHasher());
+            assertFalse(underTest.test(layerManager));
+        }
+    }
+
+    private void testAdvanceOnCount(int breakAt) {
+        Predicate<LayerManager> underTest = LayerManager.ExtendCheck.advanceOnCount(breakAt);
+        LayerManager layerManager = testingBuilder().build();
+        for (int i = 0; i < breakAt - 1; i++) {
+            assertFalse(underTest.test(layerManager), "at " + i);
+            layerManager.getTarget().merge(TestingHashers.FROM1);
+        }
+        assertTrue(underTest.test(layerManager));
+    }
+
+    @Test
+    public void testAdvanceOnCount() {
+        testAdvanceOnCount(4);
+        testAdvanceOnCount(10);
+        testAdvanceOnCount(2);
+        testAdvanceOnCount(1);
+        assertThrows(IllegalArgumentException.class, () -> LayerManager.ExtendCheck.advanceOnCount(0));
+        assertThrows(IllegalArgumentException.class, () -> LayerManager.ExtendCheck.advanceOnCount(-1));
+    }
+
+    @Test
+    public void testAdvanceOnCalculatedFull() {
+        Double maxN = shape.estimateMaxN();
+        Predicate<LayerManager> underTest = LayerManager.ExtendCheck.advanceOnSaturation(shape.estimateMaxN());
+        LayerManager layerManager = testingBuilder().build();
+        while (layerManager.getTarget().getShape().estimateN(layerManager.getTarget().cardinality()) < maxN) {
+            assertFalse(underTest.test(layerManager));
+            layerManager.getTarget().merge(TestingHashers.randomHasher());
+        }
+        assertTrue(underTest.test(layerManager));
+    }
+
+    @Test
+    public void testAdvanceOnSaturation() {
+        Double maxN = shape.estimateMaxN();
+        Predicate<LayerManager> underTest = LayerManager.ExtendCheck.advanceOnSaturation(maxN);
+        LayerManager layerManager = testingBuilder().build();
+        while (layerManager.getTarget().getShape().estimateN(layerManager.getTarget().cardinality()) < maxN) {
+            assertFalse(underTest.test(layerManager));
+            layerManager.getTarget().merge(TestingHashers.randomHasher());
```

Comment:
I was concerned by the use of random hasher. However with the shape as (17, 72) I observe only 3 or 4 hashers required to saturate the shape. Maybe add this as a note with the maxN value (2.9356) in case the shape ever changes in the future so it is obvious how to update the test.